### PR TITLE
Common - Improve file checking error messages

### PR DESCRIPTION
**Pull request by LinkIsGrim**
**When merged this pull request will:**
- Let me call people stupid when they don't read.

Example formatting:
image

image

Cases for errors are:
```
Client has additional ace addons, and compats are in additional addons // CDLC/RHS/etc loaded on client but not on server
Client has additional ace addons, and compats are not in additional addons (e.g. server has no medical loaded) // ace from different places

Version mismatch between server and client // what it says on the tin

Client has outdated addons across multiple ACE sources // conflicting install
Client has outdated addons across a single ACE source // corrupted install or bad repack
```

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by TheCandianVendingMachine** [comment]
imo the fix should come first. users won't read the full error message, and we can maybe filter out more bogus issues if the order is changed

File: addons/common/functions/fnc_checkFiles.sqf [hunk 1]
```diff
@@ -164,7 +171,30 @@ if (isMultiplayer) then {
             private _addons = GVAR(clientAddons) - GVAR(serverAddons);
 
             if (_addons isNotEqualTo []) then {
-                private _errorMsg = format ["Client/Server Addon Mismatch. Client has additional addons: %1. Server modDir: %2", _addons, GVAR(serverSource)];
```
Comment (johnb432) [inline addons/common/functions/fnc_checkFiles.sqf:167]: I'm not a fan of losing info on additional addons above 3. I thin we should make 2 error messages: The current one can be used in the UI, but the other one should contains all additional addons and be printed in full in the log.

File: addons/common/functions/fnc_checkFiles.sqf [hunk 2]
```diff
@@ -137,49 +145,113 @@ if (isMultiplayer) then {
...
+                { // Evil O(n^infinityAndBeyond) loop, can't do much about it.
+                    { // Get the real mod name for the compats we're loading
+                        private _sourceModDir = configSourceMod (_cfgPatches >> _x);
+                        if !(_sourceModDir in _defaultModDirs) then {
+                            _additionalMods pushBackUnique (_loadedModsInfo select {_x select 1 == _sourceModDir} select 0 select 0);
+                        };
+                    } forEach (getArray (_cfgPatches >> _x >> "requiredAddons"));
+                } forEach _additionalCompats;
+
+                _reasonMsg = format ["Reason: %1", "Client has extra mods requiring compats loaded (see list below)<br/><br/>"];
```
Comment (PabstMirror) [inline addons/common/functions/fnc_checkFiles.sqf:207]:
```suggestion
                private _reasonMsg = format ["Reason: %1", "Client has extra mods requiring compats loaded (see list below)<br/><br/>"];
```

File: addons/common/functions/fnc_checkFiles.sqf [hunk 3]
```diff
@@ -137,49 +145,113 @@ if (isMultiplayer) then {
...
+                _infoMsgLog
+            };
+
+            // Figure out why we have a mismatch and where it's coming from
+            // Integrated compats are the usual culprit
+            private _additionalCompats = _addons select {(_x select [0, 10]) == "ace_compat"};
+
+            if (_additionalCompats isNotEqualTo []) exitWith { // CDLC/content mod with integrated compats is loaded when it shouldn't be
+                // Fix is easy
+                _fixMsg = format ["Fix: %1", "Make sure your mod list matches. Check your server files and '-mod=' parameter if you're the server administrator."];
```
Comment (PabstMirror) [inline addons/common/functions/fnc_checkFiles.sqf:192]:
```suggestion
                private _fixMsg = format ["Fix: %1", "Make sure your mod list matches. Check your server files and '-mod=' parameter if you're the server administrator."];
```

File: addons/common/functions/fnc_checkFiles.sqf [hunk 4]
```diff
@@ -137,49 +145,113 @@ if (isMultiplayer) then {
...
+                _reasonMsg = format ["Reason: %1", "Client has extra mods requiring compats loaded (see list below)<br/><br/>"];
+                _reasonMsg = _reasonMsg + format ["Additional compatibility is being loaded for: %1", _additionalMods joinString ", "];
+
+                [_title, _reasonMsg, _fixMsg, _infoMsg, _infoMsgLog] // return
             };
 
-            private _addons = GVAR(clientAddons) - GVAR(serverAddons);
+            // Server most likely just uses a different ACE repack with some components removed
+            _reasonMsg = format ["Reason: %1", "Client and Server use ACE from different sources."];
+            _fixMsg = format ["Fix: %1", "Make sure you're using ACE from the same workshop item or repository."];
```
Comment (PabstMirror) [inline addons/common/functions/fnc_checkFiles.sqf:215]:
```suggestion
            private _reasonMsg = format ["Reason: %1", "Client and Server use ACE from different sources."];
            private _fixMsg = format ["Fix: %1", "Make sure you're using ACE from the same workshop item or repository."];
```

File: addons/common/functions/fnc_checkFiles.sqf [hunk 5]
```diff
@@ -137,51 +152,130 @@ if (isMultiplayer) then {
...
                 call FUNC(checkFiles_diagnoseACE);
 
-                ERROR(_errorMsg);
+                // Build the error message
+                // Could be a simple select but leaving the switch so it's easier to expand later
+                private _errorBuilder = switch (true) do {
+                    case _versionMismatch: {_fnc_diagnose_versionMismatch};
+                    case _addonMismatch: {_fnc_diagnose_addonMismatch};
+                };
+                (call _errorBuilder) params ["_title", "_reasonMsg", "_fixMsg", "_infoMsg", "_infoMsgLog"];
```
Comment (PabstMirror) [inline addons/common/functions/fnc_checkFiles.sqf:258]: just a possibility 
```suggestion
                (switch (true) do {
                    case _versionMismatch: _fnc_diagnose_versionMismatch;
                    case _addonMismatch: _fnc_diagnose_addonMismatch;
                }) params ["_title", "_reasonMsg", "_fixMsg", "_infoMsg", "_infoMsgLog"];
```

File: addons/common/functions/fnc_checkFiles.sqf [hunk 6]
```diff
@@ -137,51 +152,130 @@ if (isMultiplayer) then {
...
-                    ["[ACE] ERROR", _errorMsg] call FUNC(errorMessage);
+            [_title, _reasonMsg, _fixMsg, _infoMsg, _infoMsg] // return
+        };
+
+        private _fnc_diagnose_addonMismatch = {
+            private _title = "[ACE] ERROR: ADDON MISMATCH";
+
+            // Figure out why we have a mismatch and where it's coming from
+            // Integrated compats are the usual culprit
+            private _additionalCompats = _addons select {(_x select [0, 10]) == "ace_compat"};
```
Comment (johnb432) [inline addons/common/functions/fnc_checkFiles.sqf:186]: Is `_addons` guaranteed to be available in that scope?

**Review by johnb432 — APPROVED** [review]
From a logic point of view, LGTM

**Pull request merged**